File: 2021/day2/solution.py
from pathlib import Path
from dataclasses import dataclass


@dataclass
class Command:
    type: str
    qty: int


@dataclass
class State:
    x: int
    y: int
    aim: int = 0


cmd_functions_part1 = {
    'forward': lambda s, qty: State(s.x + qty, s.y),
    'down': lambda s, qty: State(s.x, s.y + qty),
    'up': lambda s, qty: State(s.x, s.y - qty)
}

cmd_functions_part2 = {
    'forward': lambda s, qty: State(s.x + qty, s.y + s.aim * qty, s.aim),
    'down': lambda s, qty: State(s.x, s.y, s.aim + qty),
    'up': lambda s, qty: State(s.x, s.y, s.aim - qty),
}


def parse_input(filename) -> list[Command]:
    def parse_line(line) -> Command:
        cmd, qty = line.split()
        return Command(cmd, int(qty))

    return [parse_line(line) for line in Path(filename).read_text().splitlines()]


def final_state(commands: list[Command], cmd_functions) -> State:
    state = State(0, 0)
    for cmd in commands:
        state = cmd_functions[cmd.type](state, cmd.qty)
    return state


commands = parse_input('input.txt')
final_pos_part1 = final_state(commands, cmd_functions_part1)
final_pos_part2 = final_state(commands, cmd_functions_part2)

assert final_pos_part1.x * final_pos_part1.y == 1480518
assert final_pos_part2.x * final_pos_part2.y == 1282809906

File: 2021/day2/solution.ts
import { PathOrFileDescriptor, readFileSync } from 'fs'
import assert from 'assert'

interface Command {
    type: string
    qty: number
}

interface State {
    x: number
    y: number
    aim: number
}

type CmdFunctions = Record<string, (s: State, qty: number) => State>

const cmdFunctionsPart1: CmdFunctions = {
    forward: (s, qty) => ({ ...s, x: s.x + qty }),
    down: (s, qty) => ({ ...s, y: s.y + qty }),
    up: (s, qty) => ({ ...s, y: s.y - qty }),
}

const cmdFunctionsPart2: CmdFunctions = {
    forward: (s, qty) => ({ ...s, x: s.x + qty, y: s.y + s.aim * qty }),
    down: (s, qty) => ({ ...s, aim: s.aim + qty }),
    up: (s, qty) => ({ ...s, aim: s.aim - qty }),
}

function parseCommands(filename: PathOrFileDescriptor): Command[] {
    function parseCommand(line: String): Command {
        const [type, qty] = line.split(' ')
        return { type, qty: parseInt(qty) }
    }

    return readFileSync(filename, 'utf-8').split('\n').map(parseCommand)
}

function finalState(commands: Command[], cmdFunctions: CmdFunctions) {
    let state = { x: 0, y: 0, aim: 0 }
    for (const cmd of commands) {
        state = cmdFunctions[cmd.type](state, cmd.qty)
    }
    return state
}

const commands = parseCommands('input.txt')

const finalPosPart1 = finalState(commands, cmdFunctionsPart1)
const finalPosPart2 = finalState(commands, cmdFunctionsPart2)

assert.equal(1480518, finalPosPart1.x * finalPosPart1.y)
assert.equal(1282809906, finalPosPart2.x * finalPosPart2.y)
